fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -2578,15 +2588,13 @@ private boolean areAllRecordTypesSince(@Nullable Collection<RecordType> recordTy
         final boolean newStore = oldFormatVersion == 0;
         final Map<Index, List<RecordType>> indexes = metaData.getIndexesSince(oldMetaDataVersion);
         if (!indexes.isEmpty()) {
-            CompletableFuture<Long> recordCountFuture = getRecordCountForRebuildIndexes(newStore, rebuildRecordCounts, indexes);
-            return recordCountFuture.thenCompose(recordCount -> {
-                if (recordCount == 0) {
+            boolean checkUpgrade = formatVersion >= SAVE_UNSPLIT_WITH_SUFFIX_FORMAT_VERSION && omitUnsplitRecordSuffix;
+            CompletableFuture<Pair<Long, Boolean>> recordCountFuture = getRecordCountForRebuildIndexes(newStore, rebuildRecordCounts, indexes, checkUpgrade);
```

Comment:
So, if I understand the problem correctly, the issue was that if there was a change to the indexes, then this would also take the opportunity to remove the format version legacy mode if it happened to be reading the count any way. (But then the record count for record stores where all primary keys begin with the type is the record count only for effected types, not for all types, which is where the problem comes from.) I suppose another strategy would be to say "forget that optimization", and then just let it keep it "omit unsplit suffix" legacy mode in that case. I think there's one unit test that would need to be removed or changed, but it might be simpler.

Reply:
I guess another middle ground would be to return `(record count, was all records)` from `getRecordCountForRebuildIndexes`. Then this would perform the upgrade if and only if `record count` was zero and `was all records` was true (which saves it from doing an extra read in the case where it wasn't reading the full record count anyway).

Reply:
I'm not opposed to any of that. Do you think it makes things significantly easier to maintain going forward?

Reply:
Hm, good question. My feeling is that it would simplify things and...probably make this part of the code base easier to maintain. It would mean that empty stores needlessly need to stay in this legacy mode, which maybe isn't the best, but it's also somewhat niche. (Like, the store would need to be empty and there would need to be an index added, all to an existing store, so it might just not be worth it.)

We could also take this as-is.

Reply:
I went with the second suggestion, as that somewhat simplifies the code and does save an extra read. It seems to me that having the record type that is adding a new index be empty is a common case and having all the other types also be empty is relatively less so, so checking every time just in case seems wasteful.